dask/order.py

```diff
@@ -381,81 +381,79 @@ def process_runnables() -> None:
-                    candidates = reachable_hull & candidates
-
-                if not candidates:
-                    if seed := pick_seed():
-                        candidates = leafs_connected[seed]
-                    else:
-                        candidates = runnable_hull or reachable_hull
-                # FIXME: This can be very expensive
-                return min(candidates, key=skey)
+    # Only build this mapping if it's needed. This can be expensive for
+    # all-to-all communications, but we normally don't need it in this
+    # case
+    occurences_grouped_sorted: dict[int, list[Key]] = {}
+
+    def _built_occurrences_group_sorted() -> None:
+        if occurences_grouped_sorted:
+            return
+        occurrences: defaultdict[Key, int] = defaultdict(int)
+        for t in leaf_nodes:
+            for r in roots_connected[t]:
+                occurrences[r] += 1
+        occurences_grouped = defaultdict(set)
+        for root, occ in occurrences.items():
+            occurences_grouped[occ].add(root)
+        occurences_grouped_sorted_tmp = {}
+        for k, v in occurences_grouped.items():
+            occurences_grouped_sorted_tmp[k] = sorted(v, key=sort_key, reverse=True)
+        occurences_grouped_sorted.update(occurences_grouped_sorted_tmp)
+        return None
+
+    def pick_seed() -> Key | None:
+        if not occurences_grouped_sorted:
+            _built_occurrences_group_sorted()
+
+        while occurences_grouped_sorted:
+            key = max(occurences_grouped_sorted)
+            picked_root = occurences_grouped_sorted[key][-1]
+            if picked_root in result:
+                occurences_grouped_sorted[key].pop()
+                if not occurences_grouped_sorted[key]:
+                    del occurences_grouped_sorted[key]
+                continue
+            return picked_root
+        return None
+
+    def get_target(longest_path: bool = False) -> Key:
+        # Some topologies benefit if the node with the most dependencies
+        # is used as first choice, others benefit from the opposite.
+        candidates = leaf_nodes
+        skey: Callable = sort_key
+
+        preferred_candidates = set()
+        if runnable_hull:
+            skey = lambda k: (num_needed[k], sort_key(k))
+            preferred_candidates = runnable_hull & candidates
+        if not preferred_candidates and reachable_hull:
+            skey = lambda k: (num_needed[k], sort_key(k))
+            preferred_candidates = reachable_hull & candidates
+
+        if reachable_hull or runnable_hull:
+            # We can't reach a leaf node directly, but we still have nodes
+            # with results in memory, these notes can inform our path towards
+            # a new preferred leaf node.
+            if not preferred_candidates:
+                hull = runnable_hull if runnable_hull else reachable_hull
+                for c in hull:
+                    preferred_candidates.update(leafs_connected[c])
```

Comment:
this looks expensive

Reply:
Can you elaborate a little? The repeated updates or the iteration generally?

Reply:
both. The hull can be possibly very large, particularly the reachable hull so we're iterating a lot here.

Particularly for dense graphs the nodes may also all be connected to all leafs so the update would handle a large set every iteration (while there are few new elements, it still has to compare all the hashes and objects)

Reply:
removed this for the alternative we are discussing below